packages/jss/tests/functional/sheet.js

```diff
@@ -636,4 +636,278 @@ describe('Functional: sheet', () => {
+    })
+
+    it('should render correct CSS', () => {
+      expect(getCss(style)).to.be(removeWhitespace(sheet.toString()))
+    })
+
+    it('should register the rule', () => {
+      expect(sheet.getRule('a')).not.to.be(rule)
+      expect(sheet.getRule('a')).to.be(replaceResult)
+    })
+
+    it('should add a rule to a detached sheet', () => {
+      sheet.detach()
+      sheet.addRule('b', {display: 'block'})
+      const newRule = sheet.replaceRule('b', {display: 'flex'})
+      sheet.attach()
+      expect(sheet.getRule('b')).to.be(newRule)
+    })
+
+    it('should link sheet in rules options', () => {
+      expect(sheet.getRule('a').options.sheet).to.be(sheet)
+    })
+
+    it('should not duplicate cssRules when adding replace to a detached sheet with link: true', () => {
+      sheet.detach()
+      sheet = jss.createStyleSheet(null, {link: true}).attach()
+      sheet.addRule('a', {float: 'left'})
+      sheet.detach()
+      sheet.addRule('b', {display: 'block'})
+      sheet.replaceRule('b', {display: 'flex'})
+      sheet.attach()
+      style = getStyle()
+      expect(getCss(style)).to.be(removeWhitespace(sheet.toString()))
+      expect(getCss(style)).to.be('.a-id{float:left;}.b-id{display:flex;}')
+    })
+  })
+
+  describe('.replaceRule() with @media and attached sheet', () => {
+    let style
+    let sheet
+
+    beforeEach(() => {
+      sheet = jss.createStyleSheet({}, {link: true}).attach()
+      sheet.addRule('a', {color: 'red'})
+      // It is important to use exactly this query, because
+      // IE will add "all" always when `cssRules.insertRule` is used,
+      // however all others will always remove "all" if query contains a second
+      // condition.
+      sheet.addRule('@media all', {
+        a: {color: 'green'}
+      })
+      sheet.replaceRule('@media all', {
+        a: {color: 'yellow'}
+      })
+      style = getStyle()
+    })
+
+    afterEach(() => {
+      sheet.detach()
+    })
+
+    it('should render @media', () => {
+      expect(getCss(style)).to.be(removeWhitespace(sheet.toString()))
+    })
+
+    it('should render @media and replaced rule', () => {
+      sheet.replaceRule('a', {
+        color: 'red'
+      })
+      expect(getCss(style)).to.be(removeWhitespace(sheet.toString()))
+    })
+  })
+
+  describe('.replaceRule() with empty styles', () => {
+    let sheet
+    let style
+
+    beforeEach(() => {
+      sheet = jss.createStyleSheet().attach()
+      sheet.addRule('a', {zIndex: 1})
+      sheet.replaceRule('a', {})
+      style = getStyle()
+    })
+
+    afterEach(() => {
+      sheet.detach()
+    })
+
+    it('should not render', () => {
+      expect(getCss(style)).to.be('')
+    })
+
+    it('should not warn', () => {
+      expect(spy.callCount).to.be(0)
+    })
+  })
+
+  describe('.replaceRule() with @keyframes and attached sheet', () => {
+    let style
+    let sheet
+
+    // We skip this test as keyframes are not supported by browser.
```

Comment:
we should drop support the support of IE9, probably even IE10 too, so this won't be needed, but also not in the same pr, just letting you know